File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return res;
    }
    
    int helper(TreeNode root) {
        if (root == null) return 0;
        
        int dl = helper(root.left);
        int dr = helper(root.right);
        
        res = Math.max(res, dl + dr);
        
        return Math.max(dl, dr) + 1;
    }
}

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def diameterOfBinaryTree(root: TreeNode): Int = {
        helper(root)._2
    }
    
    def helper(root: TreeNode): (Int, Int) = {
        if (root == null) (-1, 0)
        else {
            val (ll, ml) = helper(root.left)
            val (lr, mr) = helper(root.right)
            ((ll + 1) max (lr + 1), ml max mr max (2 + ll + lr))
        }
    }
}